### DISCUSSION TITLE
Where is it documented how a gitignore file affects the behaviour of NX?

### DISCUSSION THREAD

**Discussion by airtonix** [nrwl/nx#21126] [General]
I just wasted four hours trying to work this out, because yet again. something critical was glossed over.

> ⚠️ Preface
> 
> My existing repo had a `.gitignore` with the following: 
> 
>     *
>     !.gitignore
>     
>     !.vscode/
>     !.vscode/*
>     !.github/
>     !.github/*
>     !.github/**/*
>     !docker-compose*
>     !Taskfile.yml
>     !.editorconfig
>     !.releaserc.json
>     !yarn.lock
>     !package.json
>     
>     !tools/
>     !tools/**
>     !frontend/
>     !frontend/**

# Suggestions

1. The `.gitignore` behaviour should be very obviously mentioned in all the migration tutorials
2. The `.nxignore` page needs to say more than `.nxignore` is the same format as `.gitignore`... probably even have an equivilant page called `.gitignore`

## 1. Ran through the getting started: 

![image](https://github.com/nrwl/nx/assets/61225/e53abe9b-ffaa-4d62-86a2-091e6dd13416)

- i want to add it to an existing repo, so i ran `nx nx@latest init` ✅ 
- 🛑  first problem: something about nx failed because i didn't have `@nx/js` installs. so i add that.
- I want to manage aspects of nextjs apps, so I added `@nx/next` ✅ 
- I wanted deps to only be described at the root of the repo, so I moved them to the root package.json
- I generate a empty nextjs app with `yarn nx g @nx/next:app web` ✅ 
- I move all my old source files into this new apps `src` directory.

> 1️⃣  ❗ First and main problem
> 
> `yarn nx graph` doesn't reveal the `apps/web` project despite it having a `project.json`. 

## 2. try to work out what's broken

> ⚠️ Thoughts
> 
> at this point I'm still not aware that `.gitignore` is affecting how nx works.
> 

- generate a new blank nx monorepo
  - this one behaves as expected. 
- start copying `project.json`, `nx.json` over to the first monorepo... maybe `nx init` is broken and `npx create-nx-workspace@latest` is not. I dont know and i'm not going to spend the rest of my weekend examining source code of nx.dev to work that out.
- app is still ignored.... ❗ ❗ ❗ 

## 3. Start reading nx docs website in detail

> ⚠️ Thoughts
> 
> at this point I'm still not aware that `.gitignore` is affecting how nx works.
> 

- search for `project.json` ? did i miss something? no. everything looks normal
- search github issues for `project ignored` or `project not found`. there's one or two, but in the end not able to help my scenario.

## 4. start trying random things

- start comparing files between my monorepo and the blank working one
- come across the `.gitignore`. think, maybe this is something? nah surely not.
- comment out the above.... 🤯 🤯 🤯 🤯  it works. wtf.
- go back to searching the nx docs website for any mention of this... find nothing specifically mentioning that `.gitignore` has this much importance on the dep graph.
- come here to github discussions to rant and complain.

**Comment by shprota** [comment]
I think, nx should not ignore the projects in .gitignore. There are use cases when a developer wants to have a local playground project within the monorepo without the risk of accidentally committing it. `.nxignore` should be the only source for ignoring any projects.
